### ISSUE TITLE
Crashlytics "SetCustomKey" crash (v6.1.1)

### ISSUE THREAD

**Issue by Vah94** [firebase/quickstart-unity#385]
Unity editor version: 2018.4.0f1
Firebase Unity SDK version: 6.1.1
Firebase plugins in use (Analytics, Crashlytics):
Additional SDKs you are using (Facebook, Applovin): 
Platform you are using the Unity editor on (Mac, Windows): 
Platform you are targeting (Android): 

When i set custom key for crashlytics(Crashlytics.SetCustomKey(key, value);)

06-14 12:21:56.033 2124-2302/? E/CRASH: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
    Build type 'Release', Scripting Backend 'mono', CPU 'armeabi-v7a'
    Build fingerprint: 'samsung/dreamltexx/dreamlte:9/PPR1.180610.011/G950FXXU4DSE4:user/release-keys'
    Revision: '10'
    pid: 2124, tid: 2302, name: UnityMain  >>> com.xxxxxxxx.xxxxxxx <<<
        r0 00000000  r1 00000001  r2 b6b0a7a8  r3 b6b0a7a8
        r4 b3f03940  r5 b6b2dbc0  r6 b6b2dbc0  r7 b6b0a7a8
        r8 c3e4c674  r9 bbf79290  sl ae857bd0  fp c3e4bf80
        ip c0ca51f8  sp c3e4bf60  lr e4e4b048  pc c0c01f60  cpsr 000008fe
    backtrace:
06-14 12:21:56.036 2124-2302/? E/CRASH: 	#00  pc 000fcf60  /data/app/com.xxxxxxxx.xxxxxxx-IJ3wLlRFoNnOf8aSVS0uSg==/lib/arm/libFirebaseCppApp-6.1.1.so (firebase::crashlytics::internal::CrashlyticsInternal::SetCustomKey(char const*, char const*)+15)
    	#01  pc 00009044   ( (wrapper managed-to-native) Firebase.Crashlytics.CrashlyticsInternalPINVOKE:FirebaseCrashlyticsInternal_SetCustomKey (System.Runtime.InteropServices.HandleRef,string,string) {0xc3b57320} + 0x7c (0xe4e4afc8 0xe4e4b0a4) [0xe0bb0f00 - Unity Root Domain]+36932)
    memory near r2:
        b6b0a788 c50961b8 00180000 c50961b8 00180000  .a.......a......
        b6b0a798 c50961b8 00180000 c50961b8 00180000  .a.......a......
06-14 12:21:56.051 2124-2302/? E/MessageQueue: IdleHandler threw exception
    java.lang.Error: signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 00000000
    Build fingerprint: 'samsung/dreamltexx/dreamlte:9/PPR1.180610.011/G950FXXU4DSE4:user/release-keys'
    Revision: '10'
    pid: 2124, tid: 2302, name: UnityMain  >>> com.xxxxxxxx.xxxxxxx<<<
        r0 00000000  r1 00000001  r2 b6b0a7a8  r3 b6b0a7a8
        r4 b3f03940  r5 b6b2dbc0  r6 b6b2dbc0  r7 b6b0a7a8
        r8 c3e4c674  r9 bbf79290  sl ae857bd0  fp c3e4bf80
        ip c0ca51f8  sp c3e4bf60  lr e4e4b048  pc c0c01f60  cpsr 000008fe
    
        at libFirebaseCppApp-6.firebase::crashlytics::internal::CrashlyticsInternal::SetCustomKey(char const*, char const*)(crashlytics:15)
        at Unknown.00009044(Unknown Source:0)

**Comment by patm1987** [comment]
Hi @Vah94,

Thanks for reporting this bug. First, it appears that you're running on Android from the crash log. Can you tell me when you're calling SetCustomKey? If it does not depend on your call to `Firebase.FirebaseApp.CheckAndFixDependenciesAsync()`, can you try delaying the setting the custom key until after this completes (either in the continuation, or in a scene/prefab that loads when it completes).

If this doesn't help, can you fill out the fields at the top of the bug report (including the one I inferred, it will help anyone reading this catch up)?
Firebase plugins in use:
Additional SDKs you are using:
Platform you are using the Unity editor on:
Platform you are targeting:

In addition to this, it really helps isolate the issue better and get a fix or workaround out faster if you can come up with repro steps for the Crashlytics quickstart: https://github.com/firebase/quickstart-unity/tree/master/crashlytics/testapp

--Patrick

**Comment by Vah94** [comment]
How it used.
`using System;
using System.Collections.Generic;
using System.Linq;
using Firebase;
using UnityEngine;
using UnityEngine.Purchasing;

namespace Cats.Controllers
{
    public class FireBaseAnalyticsManager : BaseAnalyticsManager
    {
        public FireBaseAnalyticsManager()
        {
            if (FirebaseApp.CheckDependencies() != DependencyStatus.Available)
            {
                FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task => { Initialize(); });
            }
            else
            {
                Initialize();
            }
        }

        private void Initialize()
        {
            var dependencyStatus = FirebaseApp.CheckDependencies();
            if (dependencyStatus == Firebase.DependencyStatus.Available)
            {
                IsInitialized = true;
                Firebase.Analytics.FirebaseAnalytics.LogEvent(Firebase.Analytics.FirebaseAnalytics.EventAppOpen);
            }
            else
            {
                UnityEngine.Debug.LogError($"Could not resolve all Firebase dependencies: {dependencyStatus}");
            }
        }

        protected override void FireLogEvent(string eventName = "")
        {
            Firebase.Analytics.FirebaseAnalytics.LogEvent(eventName);
        }

        protected override void FireTrackPurchase(PurchaseEventArgs args)
        {
            FireLogEvent(args.purchasedProduct.definition.id);
        }

        protected override void FireHardCurrencyProfit(string eventName, double hardCurrency)
        {
            Firebase.Analytics.FirebaseAnalytics.LogEvent(
                eventName,
                Firebase.Analytics.FirebaseAnalytics.ParameterValue,
                hardCurrency);
        }

        protected override void FireLogEventWithParameter(string eventName, Dictionary<string, string> eventParams)
        {
            List<KeyValuePair<string, string>> paramList = eventParams.ToList();
            var par = new Firebase.Analytics.Parameter[paramList.Count];

            for (int i = 0; i < paramList.Count; i++)
            {
                par[i] = new Firebase.Analytics.Parameter(paramList[i].Key, paramList[i].Value);
            }

            Firebase.Analytics.FirebaseAnalytics.LogEvent(eventName, par);
        }
    }
}`

**Comment by stewartmiles** [comment]
@Vah94 the code you shared doesn't use `Crashlytics.SetCustomKey`, did you mean to paste something else?

**Comment by Vah94** [comment]
`using System.Globalization;
using Cats.UserData;
using Common.UI.Popup;
using Common.UserData;
using Firebase;
using Firebase.Crashlytics;
using UnityEngine;

namespace Cats.Controllers
{
    public class CrashlyticsManager : MonoBehaviour
    {
        private SupportManager _supportManager;
        private DataManager _dataManager;
        private PopupManager _popupManager;
        private bool _isInitialized;

        public void Init(
            bool debugCanvasActive,
            SupportManager supportManager,
            DataManager dataManager,
            PopupManager popupManager)
        {
            _supportManager = supportManager;
            _dataManager = dataManager;
            _popupManager = popupManager;
            if (!debugCanvasActive)
            {
                Application.logMessageReceivedThreaded += OnLogMessageReceived;
            }

            _isInitialized = false;
            if (FirebaseApp.CheckDependencies() != DependencyStatus.Available)
            {
                FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task => { Initialize(); });
            }
            else
            {
                Initialize();
            }
        }

        private void Initialize()
        {
            var dependencyStatus = FirebaseApp.CheckDependencies();
            if (dependencyStatus == Firebase.DependencyStatus.Available)
            {
                _isInitialized = true;
                InitKeys();
            }
            else
            {
                UnityEngine.Debug.LogError($"Could not resolve all Firebase dependencies: {dependencyStatus}");
            }
        }

        private void InitKeys()
        {
            _dataManager.AddValueChangeAction(
                UserDataValueNames.SessionCount.ToString(),
                OnSessionCountChanged);
            OnSessionCountChanged();
            _dataManager.AddValueChangeAction(
                UserDataValueNames.CompletedTutorials.ToString(),
                OnCompletedTutorialsChanged);
            OnCompletedTutorialsChanged();
            _dataManager.AddValueChangeAction(
                UserDataValueNames.UnlockLevel.ToString(),
                OnUnlockLevelChanged);
            OnUnlockLevelChanged();
            _popupManager.OnAnyPopupShownChanged += popupType =>
            {
                if (popupType != null)
                {
                    SetCustomKeys("Last popup", popupType.ToString());
                }
            };

            _dataManager.AddValueChangeAction(
                UserDataValueNames.SoftCurrency.ToString(),
                OnSoftCurrencyChanged);
            OnSoftCurrencyChanged();
            _dataManager.AddValueChangeAction(
                UserDataValueNames.HardCurrency.ToString(),
                OnHardCurrencyChanged);
            OnHardCurrencyChanged();
        }

        private void OnHardCurrencyChanged()
        {
            SetCustomKeys(
                UserDataValueNames.HardCurrency.ToString(),
                _dataManager.HardCurrency.Value.ToString(CultureInfo.InvariantCulture));
        }

        private void OnSoftCurrencyChanged()
        {
            SetCustomKeys(
                UserDataValueNames.SoftCurrency.ToString(),
                _dataManager.SoftCurrency.Value.ToString(CultureInfo.InvariantCulture));
        }

        private void OnUnlockLevelChanged()
        {
            SetCustomKeys(
                UserDataValueNames.UnlockLevel.ToString(),
                _dataManager.UnlockLevel.ToString(CultureInfo.InvariantCulture));
        }

        private void OnCompletedTutorialsChanged()
        {
            var completedTutorials = "";
            foreach (var item in _dataManager.CompletedTutorials)
            {
                completedTutorials += ((StringUserValue) item).GetValue();
            }

            SetCustomKeys(UserDataValueNames.CompletedTutorials.ToString(), completedTutorials);
        }

        private void OnSessionCountChanged()
        {
            SetCustomKeys(
                UserDataValueNames.SessionCount.ToString(),
                _dataManager.SessionCount.ToString(CultureInfo.InvariantCulture));
        }

        public void OnLocationChanged(string value)
        {
            SetCustomKeys("Location", value);
        }

        private void OnLogMessageReceived(string logString, string stackTrace, LogType type)
        {
            if (type == LogType.Exception && !Application.isEditor)
            {
                Application.logMessageReceivedThreaded -= OnLogMessageReceived;
                _supportManager?.SendMailToSupport(logString);
                Application.Quit();
            }
        }

        private void SetCustomKeys(string key, string value)
        {
            if (!_isInitialized)
            {
                return;
            }

            if (Application.platform == RuntimePlatform.Android)
            {
                //TODO Crashlytics 6.1.1 is not work SetCustomKey function... waiting for new version
                return;
            }

            Crashlytics.SetCustomKey(key, value);
        }
    }
}`

**Comment by stewartmiles** [comment]
@Vah94 I spent some time trying to reproduce the failure you're seeing with no luck.  However, while tracking down the cause of https://github.com/firebase/quickstart-unity/issues/392 the fix will likely impact what you're seeing as well.

So that we can understand how you're getting into this state could you share the output of "adb logcat" for a session that crashes?

e.g on macOS `adb logcat -c; adb logcat | tee logcat.log`

**Comment by stewartmiles** [comment]
@Vah94 thanks for the log, are you able to reproduce the same behavior in our sample application?  The first thing that worries me about your log are lines 1379...1420

```
06-27 16:57:22.073 26675-26762/? I/firebase: Firebase App initializing app com.company.app (default 1).
06-27 16:57:22.074 26675-26762/? I/Unity: Firebase App initializing app com.company.app (default 1).
     
    (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
06-27 16:57:22.074 26675-26762/? I/firebase: Firebase Analytics API Initializing
06-27 16:57:22.075 26675-26762/? I/Unity: Firebase Analytics API Initializing
     
    (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
06-27 16:57:22.075 26675-26762/? I/firebase: analytics API Initialized
06-27 16:57:22.076 26675-26762/? I/Unity: analytics API Initialized
     
    (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
06-27 16:57:22.085 26675-26762/? W/firebase: Crashlytics object 0x209528c8 should be deleted before the App 0x251ac480 it depends upon.
06-27 16:57:22.086 26675-26762/? W/Unity: Crashlytics object 0x209528c8 should be deleted before the App 0x251ac480 it depends upon.
     
    (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
06-27 16:57:22.089 26675-26762/? I/games.app: The ClassLoaderContext is a special shared library.
06-27 16:57:22.093 26675-26762/? I/games.app: The ClassLoaderContext is a special shared library.
06-27 16:57:22.094 26675-26762/? I/firebase: Firebase App initializing app com.company.app (default 1).
06-27 16:57:22.095 26675-26762/? I/Unity: Firebase App initializing app com.company.app (default 1).
     
    (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
06-27 16:57:22.096 26675-26762/? I/firebase: Firebase Analytics API Initializing
06-27 16:57:22.097 26675-26762/? I/Unity: Firebase Analytics API Initializing
     
    (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
06-27 16:57:22.097 26675-26762/? I/firebase: analytics API Initialized
06-27 16:57:22.098 26675-26762/? I/Unity: analytics API Initialized
     
    (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
06-27 16:57:22.103 26675-26762/? I/games.app: The ClassLoaderContext is a special shared library.
06-27 16:57:22.106 26675-26762/? I/games.app: The ClassLoaderContext is a special shared library.
06-27 16:57:22.108 26675-26762/? I/firebase: Firebase App initializing app com.company.app (default 1).
06-27 16:57:22.109 26675-26762/? I/Unity: Firebase App initializing app com.company.app (default 1).
     
    (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
06-27 16:57:22.109 26675-26762/? I/firebase: Firebase Analytics API Initializing
06-27 16:57:22.110 26675-26762/? I/Unity: Firebase Analytics API Initializing
     
    (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
06-27 16:57:22.110 26675-26762/? I/firebase: analytics API Initialized
06-27 16:57:22.111 26675-26762/? I/Unity: analytics API Initialized
```
It looks like everything is being initialized twice which makes me wonder what else you're doing in your application that could be causing this.  Is it possible you're running two instances of FIrebase?  Do you have two libFirebaseCppApp*.so in your APK, e.g you should see one per architecture...

```
$ unzip -l unity_test/imported/dotnet4/crashlytics/testapp/a.apk | grep libFirebase
  2868244  00-00-1980 00:00   lib/armeabi-v7a/libFirebaseCppApp-6.1.1.so
    40560  00-00-1980 00:00   lib/armeabi-v7a/libFirebaseCppCrashlytics.so
  3614412  00-00-1980 00:00   lib/x86/libFirebaseCppApp-6.1.1.so
    64656  00-00-1980 00:00   lib/x86/libFirebaseCppCrashlytics.so
```

Also, it looks like you're using IL2CPP so I enabled it and tried hacking our sample application (see attached patched file) to test what could be happening in your application .. still doesn't fail.

[UIHandler.cs.zip](https://github.com/firebase/quickstart-unity/files/3335563/UIHandler.cs.zip)

**Comment by Vah94** [comment]
[Firebase.txt](https://github.com/firebase/quickstart-unity/files/3339430/Firebase.txt)
[AllLogs.txt](https://github.com/firebase/quickstart-unity/files/3339431/AllLogs.txt)

**Comment by stewartmiles** [comment]
@Vah94 are the logs you shared reproduced with our sample application?  We really need a minimal reproducible case for this.  I can see it's crashing in `Firebase.txt` but we really need to know what else you're doing since as I mentioned in https://github.com/firebase/quickstart-unity/issues/385#issuecomment-506408500 this isn't, as far as I can tell, reproducible.

**Comment by stewartmiles** [comment]
@Vah94 it looks like we have code in SetCustomKey that passes the C# string arguments through a PINVOKE without any null checking.  I'm not sure why we're seeing these marshalled as zero length strings in our tests.  We've added some checking in there as well to make sure both arguments of SetCustomKey can't be null.  Is it possible that one of the strings you're passing to the method is null?

**Comment by alexames** [comment]
This should be fixed as of the latest release, 6.2.0
